### ISSUE TITLE
Enable Link-Time Optimization (LTO)

### ISSUE THREAD

**Issue by zamazan4ik** [hikari-no-yume/touchHLE#371]
### Describe the feature or issue.

Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

Thank you.

### By submitting this issue, I certify that…

- [X] I understand that linking to, discussing or asking for pirated content is not permitted, and will lead to an immediate ban.
- [X] I understand that “Abandonware” and illegal copies of free games are also pirated content.

**Comment by hikari-no-yume** [comment]
This only impacts interactions between crates — all the same optimisations will be done *within* crates, right? If that's the case, I think it's unlikely this would have a significant impact on touchHLE performance or code size. The project is almost entirely within a single crate with minimal and carefully controlled use of other crates.

I guess it's worth looking into though.

**Comment by zamazan4ik** [comment]
> This only impacts interactions between crates — all the same optimisations will be done within crates, right?

Yes. However, the optimizations between crates can also lead to different optimization decisions inside a crate if a context from another crate brings to the compiler optimizer valuable information.

I have made quick tests on my local machine (Fedora 40, Rustc 1.81, the latest `touchHLE` sources, build command - `cargo build --release --bin touchHLE --no-default-features`). The results are the following:
```
-rwxr-xr-x. 1 zamazan4ik zamazan4ik_new 12108776 окт  9 19:39 touchHLE_lto
-rwxr-xr-x. 1 zamazan4ik zamazan4ik_new 12589560 окт  9 19:37 touchHLE_release
```
, where `touchHLE` - the Release binary with `lto = true` enabled.
So we see some binary size reduction but not too huge. Regarding performance perspectives - it need to be compared separately.

**Comment by hikari-no-yume** [comment]
How much did compilation time change?

**Comment by zamazan4ik** [comment]
On AMD Ryzen 9 5900x I have the following build times:
* Release: 35s
* Fat LTO: 53s

Didn't test ThinLTO though.
